lib/import.js

```diff
@@ -95,21 +128,68 @@ module.exports = (self) => {
       }
     },
 
-    async importPieces (req, pieces, reporting) {
-      let importedCount = 0;
+    async importOrUpdatePieces (req, {
+      pieces, reporting, updateKey, updateField
+    }) {
+      const counter = {
+        imported: 0,
+        updated: 0,
+        failed: 0
+      };
 
       for (const piece of pieces) {
         try {
-          await self.insert(req, piece);
+          const updateKeyValue = updateKey && piece[updateKey];
+
+          if (!updateKeyValue) {
+            const converted = {};
+
+            await self.convert(req, piece, converted);
+            await self.insert(req, converted);
 
+            counter.imported++;
+            reporting.success();
+            continue;
+          }
+
+          const existingPiece = await self.findForEditing(req, {
+            aposMode: 'draft',
+            [updateField]: updateKeyValue
+          }).toObject();
+
+          if (!existingPiece) {
+            throw new Error('No piece to update');
```

Comment:
This is not enough information to figure out which row is bad. Not a big deal to fix, just make sure you include updateKeyValue and the label of the key field and that this winds up reaching the user as a notification.

Reply:
If you need to you can use `self.apos.notify(req, ...)` server side.

Reply:
Notifications improved with more infos like for convert errors.